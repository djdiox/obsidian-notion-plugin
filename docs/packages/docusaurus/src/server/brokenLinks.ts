/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import path from 'path';
import _ from 'lodash';
import logger from '@docusaurus/logger';
import combinePromises from 'combine-promises';
import {matchRoutes} from 'react-router-config';
import {removePrefix, removeSuffix, resolvePathname} from '@docusaurus/utils';
import {getAllFinalRoutes} from './utils';
import type {RouteConfig, ReportingSeverity} from '@docusaurus/types';

type BrokenLink = {
  link: string;
  resolvedLink: string;
};

// matchRoutes does not support qs/anchors, so we remove it!
function onlyPathname(link: string) {
  return link.split('#')[0]!.split('?')[0]!;
}

function getPageBrokenLinks({
  pagePath,
  pageLinks,
  routes,
}: {
  pagePath: string;
  pageLinks: string[];
  routes: RouteConfig[];
}): BrokenLink[] {
  // ReactRouter is able to support links like ./../somePath but `matchRoutes`
  // does not do this resolution internally. We must resolve the links before
  // using `matchRoutes`. `resolvePathname` is used internally by React Router
  function resolveLink(link: string) {
    const resolvedLink = resolvePathname(onlyPathname(link), pagePath);
    return {link, resolvedLink};
  }

  function isBrokenLink(link: string) {
    const matchedRoutes = [link, decodeURI(link)]
      // @ts-expect-error: React router types RouteConfig with an actual React
      // component, but we load route components with string paths.
      // We don't actually access component here, so it's fine.
      .map((l) => matchRoutes(routes, l))
      .flat();
    return matchedRoutes.length === 0;
  }

  return pageLinks.map(resolveLink).filter((l) => isBrokenLink(l.resolvedLink));
}

/**
 * The route defs can be recursive, and have a parent match-all route. We don't
 * want to match broken links like /docs/brokenLink against /docs/*. For this
 * reason, we only consider the "final routes" that do not have subroutes.
 * We also need to remove the match-all 404 route
 */
function filterIntermediateRoutes(routesInput: RouteConfig[]): RouteConfig[] {
  const routesWithout404 = routesInput.filter((route) => route.path !== '*');
  return getAllFinalRoutes(routesWithout404);
}

function getAllBrokenLinks({
  allCollectedLinks,
  routes,
}: {
  allCollectedLinks: {[location: string]: string[]};
  routes: RouteConfig[];
}): {[location: string]: BrokenLink[]} {
  const filteredRoutes = filterIntermediateRoutes(routes);

  const allBrokenLinks = _.mapValues(allCollectedLinks, (pageLinks, pagePath) =>
    getPageBrokenLinks({pageLinks, pagePath, routes: filteredRoutes}),
  );

  return _.pickBy(allBrokenLinks, (brokenLinks) => brokenLinks.length > 0);
}

function getBrokenLinksErrorMessage(allBrokenLinks: {
  [location: string]: BrokenLink[];
}): string | undefined {
  if (Object.keys(allBrokenLinks).length === 0) {
    return undefined;
  }

  function brokenLinkMessage(brokenLink: BrokenLink): string {
    const showResolvedLink = brokenLink.link !== brokenLink.resolvedLink;
    return `${brokenLink.link}${
      showResolvedLink ? ` (resolved as: ${brokenLink.resolvedLink})` : ''
    }`;
  }

  function pageBrokenLinksMessage(
    pagePath: string,
    brokenLinks: BrokenLink[],
  ): string {
    return `
- On source page path = ${pagePath}:
   -> linking to ${brokenLinks
     .map(brokenLinkMessage)
     .join('\n   -> linking to ')}`;
  }

  /**
   * If there's a broken link appearing very often, it is probably a broken link
   * on the layout. Add an additional message in such case to help user figure
   * this out. See https://github.com/facebook/docusaurus/issues/3567#issuecomment-706973805
   */
  function getLayoutBrokenLinksHelpMessage() {
    const flatList = Object.entries(allBrokenLinks).flatMap(
      ([pagePage, brokenLinks]) =>
        brokenLinks.map((brokenLink) => ({pagePage, brokenLink})),
    );

    const countedBrokenLinks = _.countBy(
      flatList,
      (item) => item.brokenLink.link,
    );

    const FrequencyThreshold = 5; // Is this a good value?
    const frequentLinks = Object.entries(countedBrokenLinks)
      .filter(([, count]) => count >= FrequencyThreshold)
      .map(([link]) => link);

    if (frequentLinks.length === 0) {
      return '';
    }

    return logger.interpolate`

It looks like some of the broken links we found appear in many pages of your site.
Maybe those broken links appear on all pages through your site layout?
We recommend that you check your theme configuration for such links (particularly, theme navbar and footer).
Frequent broken links are linking to:${frequentLinks}`;
  }

  return `Docusaurus found broken links!

Please check the pages of your site in the list below, and make sure you don't reference any path that does not exist.
Note: it's possible to ignore broken links with the 'onBrokenLinks' Docusaurus configuration, and let the build pass.${getLayoutBrokenLinksHelpMessage()}

Exhaustive list of all broken links found:
${Object.entries(allBrokenLinks)
  .map(([pagePath, brokenLinks]) =>
    pageBrokenLinksMessage(pagePath, brokenLinks),
  )
  .join('\n')}
`;
}

async function isExistingFile(filePath: string) {
  try {
    return (await fs.stat(filePath)).isFile();
  } catch {
    return false;
  }
}

// If a file actually exist on the file system, we know the link is valid
// even if docusaurus does not know about this file, so we don't report it
async function filterExistingFileLinks({
  baseUrl,
  outDir,
  allCollectedLinks,
}: {
  baseUrl: string;
  outDir: string;
  allCollectedLinks: {[location: string]: string[]};
}): Promise<{[location: string]: string[]}> {
  async function linkFileExists(link: string) {
    // /baseUrl/javadoc/ -> /outDir/javadoc
    const baseFilePath = onlyPathname(
      removeSuffix(`${outDir}/${removePrefix(link, baseUrl)}`, '/'),
    );

    // -> /outDir/javadoc
    // -> /outDir/javadoc.html
    // -> /outDir/javadoc/index.html
    const filePathsToTry: string[] = [baseFilePath];
    if (!path.extname(baseFilePath)) {
      filePathsToTry.push(
        `${baseFilePath}.html`,
        path.join(baseFilePath, 'index.html'),
      );
    }

    for (const file of filePathsToTry) {
      if (await isExistingFile(file)) {
        return true;
      }
    }
    return false;
  }

  return combinePromises(
    _.mapValues(allCollectedLinks, async (links) =>
      (
        await Promise.all(
          links.map(async (link) => ((await linkFileExists(link)) ? '' : link)),
        )
      ).filter(Boolean),
    ),
  );
}

export async function handleBrokenLinks({
  allCollectedLinks,
  onBrokenLinks,
  routes,
  baseUrl,
  outDir,
}: {
  allCollectedLinks: {[location: string]: string[]};
  onBrokenLinks: ReportingSeverity;
  routes: RouteConfig[];
  baseUrl: string;
  outDir: string;
}): Promise<void> {
  if (onBrokenLinks === 'ignore') {
    return;
  }

  // If we link to a file like /myFile.zip, and the file actually exist for the
  // file system. It is not a broken link, it may simply be a link to an
  // existing static file...
  const allCollectedLinksFiltered = await filterExistingFileLinks({
    allCollectedLinks,
    baseUrl,
    outDir,
  });

  const allBrokenLinks = getAllBrokenLinks({
    allCollectedLinks: allCollectedLinksFiltered,
    routes,
  });

  const errorMessage = getBrokenLinksErrorMessage(allBrokenLinks);
  if (errorMessage) {
    logger.report(onBrokenLinks)(errorMessage);
  }
}
