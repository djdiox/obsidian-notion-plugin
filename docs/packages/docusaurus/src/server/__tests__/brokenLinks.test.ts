/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {jest} from '@jest/globals';
import path from 'path';
import _ from 'lodash';
import {handleBrokenLinks} from '../brokenLinks';
import type {RouteConfig} from '@docusaurus/types';

describe('handleBrokenLinks', () => {
  const routes: RouteConfig[] = [
    {
      path: '/community',
      component: '',
    },
    {
      path: '/docs',
      component: '',
      routes: [
        {path: '/docs/goodDoc', component: ''},
        {path: '/docs/anotherGoodDoc', component: ''},
        {path: '/docs/good doc with space', component: ''},
        {path: '/docs/another good doc with space', component: ''},
        {path: '/docs/weird%20but%20good', component: ''},
      ],
    },
    {
      path: '*',
      component: '',
    },
  ];

  const link1 = '/link1';
  const link2 = '/docs/link2';
  const link3 = '/hey/link3';

  const linkToJavadoc1 = '/javadoc';
  const linkToJavadoc2 = '/javadoc/';
  const linkToJavadoc3 = '/javadoc/index.html';
  const linkToJavadoc4 = '/javadoc/index.html#foo';

  const linkToZipFile = '/files/file.zip';
  const linkToHtmlFile1 = '/files/hey.html';
  const linkToHtmlFile2 = '/files/hey';

  const linkToEmptyFolder1 = '/emptyFolder';
  const linkToEmptyFolder2 = '/emptyFolder/';
  const allCollectedLinks = {
    '/docs/good doc with space': [
      // Good - valid file with spaces in name
      './another%20good%20doc%20with%20space',
      // Good - valid file with percent-20 in its name
      './weird%20but%20good',
      // Bad - non-existent file with spaces in name
      './some%20other%20non-existent%20doc1',
      // Evil - trying to use ../../ but '/' won't get decoded
      // cSpell:ignore Fout
      './break%2F..%2F..%2Fout2',
    ],
    '/docs/goodDoc': [
      // Good links
      './anotherGoodDoc#someHash',
      '/docs/anotherGoodDoc?someQueryString=true#someHash',
      '../docs/anotherGoodDoc?someQueryString=true',
      '../docs/anotherGoodDoc#someHash',
      // Bad links
      '../anotherGoodDoc#reported-because-of-bad-relative-path1',
      './docThatDoesNotExist2',
      './badRelativeLink3',
      '../badRelativeLink4',
    ],
    '/community': [
      // Good links
      '/docs/goodDoc',
      '/docs/anotherGoodDoc#someHash',
      './docs/goodDoc#someHash',
      './docs/anotherGoodDoc',
      // Bad links
      '/someNonExistentDoc1',
      '/badLink2',
      './badLink3',
    ],
    '/page1': [
      link1,
      linkToHtmlFile1,
      linkToJavadoc1,
      linkToHtmlFile2,
      linkToJavadoc3,
      linkToJavadoc4,
      linkToEmptyFolder1, // Not filtered!
    ],
    '/page2': [
      link2,
      linkToEmptyFolder2, // Not filtered!
      linkToJavadoc2,
      link3,
      linkToJavadoc3,
      linkToZipFile,
    ],
  };

  const outDir = path.resolve(__dirname, '__fixtures__/brokenLinks/outDir');

  it('do not report anything for correct paths', async () => {
    const consoleMock = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    const allCollectedCorrectLinks = {
      '/docs/good doc with space': [
        './another%20good%20doc%20with%20space',
        './weird%20but%20good',
      ],
      '/docs/goodDoc': [
        './anotherGoodDoc#someHash',
        '/docs/anotherGoodDoc?someQueryString=true#someHash',
        '../docs/anotherGoodDoc?someQueryString=true',
        '../docs/anotherGoodDoc#someHash',
      ],
      '/community': [
        '/docs/goodDoc',
        '/docs/anotherGoodDoc#someHash',
        './docs/goodDoc#someHash',
        './docs/anotherGoodDoc',
      ],
      '/page1': [
        linkToHtmlFile1,
        linkToJavadoc1,
        linkToHtmlFile2,
        linkToJavadoc3,
        linkToJavadoc4,
      ],
    };
    await handleBrokenLinks({
      allCollectedLinks: allCollectedCorrectLinks,
      onBrokenLinks: 'error',
      routes,
      baseUrl: '/',
      outDir,
    });
    expect(consoleMock).toBeCalledTimes(0);
  });

  it('reports all broken links', async () => {
    await expect(() =>
      handleBrokenLinks({
        allCollectedLinks,
        onBrokenLinks: 'throw',
        routes,
        baseUrl: '/',
        outDir,
      }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('no-op for ignore', async () => {
    // In any case, _.mapValues will always be called, unless handleBrokenLinks
    // has already bailed
    const lodashMock = jest.spyOn(_, 'mapValues');
    await handleBrokenLinks({
      allCollectedLinks,
      onBrokenLinks: 'ignore',
      routes,
      baseUrl: '/',
      outDir,
    });
    expect(lodashMock).toBeCalledTimes(0);
    lodashMock.mockRestore();
  });

  it('reports frequent broken links', async () => {
    Object.values(allCollectedLinks).forEach((links) =>
      links.push(
        '/frequent',
        // This is in the gray area of what should be reported. Relative paths
        // may be resolved to different slugs on different locations. But if
        // this comes from a layout link, it should be reported anyways
        './maybe-not',
      ),
    );

    await expect(() =>
      handleBrokenLinks({
        allCollectedLinks,
        onBrokenLinks: 'throw',
        routes,
        baseUrl: '/',
        outDir,
      }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
