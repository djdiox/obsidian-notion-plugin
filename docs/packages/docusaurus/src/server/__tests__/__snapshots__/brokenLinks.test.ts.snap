// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleBrokenLinks reports all broken links 1`] = `
"Docusaurus found broken links!

Please check the pages of your site in the list below, and make sure you don't reference any path that does not exist.
Note: it's possible to ignore broken links with the 'onBrokenLinks' Docusaurus configuration, and let the build pass.

Exhaustive list of all broken links found:

- On source page path = /docs/good doc with space:
   -> linking to ./some%20other%20non-existent%20doc1 (resolved as: /docs/some%20other%20non-existent%20doc1)
   -> linking to ./break%2F..%2F..%2Fout2 (resolved as: /docs/break%2F..%2F..%2Fout2)

- On source page path = /docs/goodDoc:
   -> linking to ../anotherGoodDoc#reported-because-of-bad-relative-path1 (resolved as: /anotherGoodDoc)
   -> linking to ./docThatDoesNotExist2 (resolved as: /docs/docThatDoesNotExist2)
   -> linking to ./badRelativeLink3 (resolved as: /docs/badRelativeLink3)
   -> linking to ../badRelativeLink4 (resolved as: /badRelativeLink4)

- On source page path = /community:
   -> linking to /someNonExistentDoc1
   -> linking to /badLink2
   -> linking to ./badLink3 (resolved as: /badLink3)

- On source page path = /page1:
   -> linking to /link1
   -> linking to /emptyFolder

- On source page path = /page2:
   -> linking to /docs/link2
   -> linking to /emptyFolder/
   -> linking to /hey/link3
"
`;

exports[`handleBrokenLinks reports frequent broken links 1`] = `
"Docusaurus found broken links!

Please check the pages of your site in the list below, and make sure you don't reference any path that does not exist.
Note: it's possible to ignore broken links with the 'onBrokenLinks' Docusaurus configuration, and let the build pass.

It looks like some of the broken links we found appear in many pages of your site.
Maybe those broken links appear on all pages through your site layout?
We recommend that you check your theme configuration for such links (particularly, theme navbar and footer).
Frequent broken links are linking to:
- /frequent
- ./maybe-not

Exhaustive list of all broken links found:

- On source page path = /docs/good doc with space:
   -> linking to ./some%20other%20non-existent%20doc1 (resolved as: /docs/some%20other%20non-existent%20doc1)
   -> linking to ./break%2F..%2F..%2Fout2 (resolved as: /docs/break%2F..%2F..%2Fout2)
   -> linking to /frequent
   -> linking to ./maybe-not (resolved as: /docs/maybe-not)

- On source page path = /docs/goodDoc:
   -> linking to ../anotherGoodDoc#reported-because-of-bad-relative-path1 (resolved as: /anotherGoodDoc)
   -> linking to ./docThatDoesNotExist2 (resolved as: /docs/docThatDoesNotExist2)
   -> linking to ./badRelativeLink3 (resolved as: /docs/badRelativeLink3)
   -> linking to ../badRelativeLink4 (resolved as: /badRelativeLink4)
   -> linking to /frequent
   -> linking to ./maybe-not (resolved as: /docs/maybe-not)

- On source page path = /community:
   -> linking to /someNonExistentDoc1
   -> linking to /badLink2
   -> linking to ./badLink3 (resolved as: /badLink3)
   -> linking to /frequent
   -> linking to ./maybe-not (resolved as: /maybe-not)

- On source page path = /page1:
   -> linking to /link1
   -> linking to /emptyFolder
   -> linking to /frequent
   -> linking to ./maybe-not (resolved as: /maybe-not)

- On source page path = /page2:
   -> linking to /docs/link2
   -> linking to /emptyFolder/
   -> linking to /hey/link3
   -> linking to /frequent
   -> linking to ./maybe-not (resolved as: /maybe-not)
"
`;
